xdsl/dialects/csl.py

```diff
@@ -344,20 +404,121 @@ def parse(cls, parser: Parser) -> FuncOp:
         return func
 
     def print(self, printer: Printer):
-        print_func_op_like(
-            printer,
-            self.sym_name,
-            self.function_type,
-            self.body,
-            self.attributes,
-            arg_attrs=self.arg_attrs,
-            reserved_attr_names=(
-                "sym_name",
-                "function_type",
-                "sym_visibility",
-                "arg_attrs",
-            ),
+        FuncBase._print(self, printer)
+
+
+@irdl_op_definition
+class TaskOp(IRDLOperation, FuncBase):
+    """
+    Represents a task in CSL. All three types of task are represented by this Op.
+
+    It carries the ID it should be bound to, in case of local and control tasks
+    this is the task ID, in the case of the data task, it's the id of the color
+    the task is bound to.
+
+    NOTE: Control tasks not yet implemented
+    """
+
+    name = "csl.task"
+
+    kind = prop_def(TaskKindAttr)
+    id = prop_def(IntegerAttr[IntegerType])
+
+    traits = frozenset([InModuleKind(ModuleKind.PROGRAM)])
+
+    def __init__(
+        self,
+        name: str,
+        function_type: FunctionType | tuple[Sequence[Attribute], Attribute | None],
+        region: Region | type[Region.DEFAULT] = Region.DEFAULT,
+        *,
+        task_kind: TaskKindAttr | TaskKind,
+        arg_attrs: ArrayAttr[DictionaryAttr] | None = None,
+        res_attrs: ArrayAttr[DictionaryAttr] | None = None,
+        id: IntegerAttr[IntegerType] | int,
+    ):
+        properties, region = self._props_region(
+            name, function_type, region, arg_attrs=arg_attrs, res_attrs=res_attrs
+        )
+        if isinstance(task_kind, TaskKind):
+            task_kind = TaskKindAttr(task_kind)
+        if isinstance(id, int):
+            id = IntegerAttr.from_int_and_width(id, task_kind.get_color_bits())
+        assert (
+            id.type.width.data == task_kind.get_color_bits()
+        ), f"{task_kind.data.value} task id has to have {task_kind.get_color_bits()} bits, got {id.type.width.data}"
+
+        properties |= {
+            "kind": task_kind,
+            "id": id,
+        }
+        super().__init__(properties=properties, regions=[region])
+
+    def verify_(self) -> None:
+        FuncBase._verify(self)
+        if len(self.function_type.outputs.data) != 0:
+            raise VerifyException(f"{self.name} cannot have return values")
+
+        # TODO(dk949): Need to check at some point that we're not reusing the same color multiple times
```

Comment:
I think we could do this as part of a pass instead? Maybe have some pass that allocates ids to tasks?

Reply:
This seems more sensible than doing it in the `verify`. Should I remove the TODO?

Reply:
In general, we can have a pass called `csl-legalize` that looks for broken programs, and tries to fix them (maybe inserting foreign function definitions, assigning task IDs or something like that) and if it can't fix an issue it bails out. Basically as a sanity check that we will print valid CSL code.